File: modules/s3/src/core2/s3/S3Configurator.java
package core2.s3;

import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;

public interface S3Configurator {
    default PutObjectRequest.Builder configurePut(PutObjectRequest.Builder builder) {
        return builder;
    }

    default GetObjectRequest.Builder configureGet(GetObjectRequest.Builder builder) {
        return builder;
    }

    default S3AsyncClient makeClient() {
        return S3AsyncClient.create();
    }
}

File: modules/aws/src/main/kotlin/xtdb/aws/s3/S3Configurator.kt
package xtdb.aws.s3

import software.amazon.awssdk.services.s3.S3AsyncClientBuilder
import software.amazon.awssdk.services.s3.model.GetObjectRequest
import software.amazon.awssdk.services.s3.model.HeadObjectRequest
import software.amazon.awssdk.services.s3.model.PutObjectRequest

interface S3Configurator {
    fun configureGet(builder: GetObjectRequest.Builder) = Unit
    fun configureHead(builder: HeadObjectRequest.Builder) = Unit
    fun configurePut(builder: PutObjectRequest.Builder) = Unit
    fun configureClient(builder: S3AsyncClientBuilder) = Unit

    data object Default: S3Configurator
}
